crypto/key.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crypto
+
+import (
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"golang.org/x/exp/slices"
+)
+
+// Key extends jwk.Key adding optional properties for determining if the key is valid (time bounds) or can be used for certain purposes.
+type Key struct {
+	jwk.Key
+
+	kid string
+	exp *time.Time
+	nbf *time.Time
+}
+
+// NewKey returns a new Key object
+func NewKey(key jwk.Key, kid string, exp, nbf *time.Time) *Key {
+	return &Key{
+		Key: key,
+		kid: kid,
+		exp: exp,
+		nbf: nbf,
+	}
+}
+
+// KeyID returns the value of the kid (key ID) property if present.
+func (k Key) KeyID() string {
+	return k.kid
+}
+
+// CanPerformOperation returns true if the key can be used to perform a specific operation.
+func (k Key) CanPerformOperation(op jwk.KeyOperation) bool {
+	return KeyCanPerformOperation(k, op)
+}
+
+// IsValid checks if the key is within the time bounds of validity.
+func (k Key) IsValid() bool {
+	return k.isValidAtTime(time.Now())
+}
+
+func (k Key) isValidAtTime(t time.Time) bool {
+	if k.exp != nil && k.exp.Before(t) {
+		return false
+	}
+
+	if k.nbf != nil && k.nbf.After(t) {
+		return false
+	}
+
+	return true
+}
+
+// KeyCanPerformOperation returns true if the key can be used to perform a specific operation.
+func KeyCanPerformOperation(key jwk.Key, op jwk.KeyOperation) bool {
```

Comment:
does this need to be an exported function? 
Can we use the above function `CanPerformOperation` on a `jwk.Key` instead?

Reply:
Yes there is a reason for this (and `KeyCanPerformAlgorithm` below) to be exported functions.

These functions need to be able to support any object that implements `jwk.Key`. The `crypto.Key` struct defined in this package implements `jwk.Key`, but extends it too. Some components do not have a `crypto.Key`, but only a `jwk.Key`.